### ISSUE TITLE
Configurable full battery capacity

### ISSUE THREAD

**Issue by nanomad** [SAIC-iSmart-API/saic-python-mqtt-gateway#72]
Referencing #69, I think @Tonno87 made I good point.

We currently have two topics that deal with kwh and they get the info from the car internal BMS which we know is wrong (always reporting a 72kWh full battery capacity with a soc_kwh computed from there).

I was thinking we can 
- Supply an internal mapping from known model/series to full kwh
- Allow the user to specify their own value with a mapping like they do for ABRP api keys
Thoughts?

**Comment by tosate** [comment]
If we croud-source this a bit like @tisoft did for the car configurations, we should get quickly the data for an internal mapping.
The manual specification is kind of a fallback for cars which are not part of the internal mapping.

I would prefer the internal mapping because it doesn't require additional configuration. I am starting and stopping my gateway quite often. Since it has no persistence, I would need to publish my own value every time.

In fact, we could use the MQTT broker for persistence with retained messages. However, retained commands lead to an error in the current gateway implementation.

**Comment by nanomad** [comment]
Yeah, I guess so. I'd much prefer the internal mapping as well with maybe a fallback just in case

**Comment by Tonno87** [comment]
I think internal mapping with a fallback is a good idea. 

So we need a value to determite the right model and battery. 
After a quick look into the [table](https://github.com/orgs/SAIC-iSmart-API/discussions/4) i found out the following and inserted the capacity from mgmotors.de:

MG5 SR  | Battery=2  50,3kWh
MG5 LR  | Battery=1  61,1kWh

MG4 S  | Battery=2  51kWh
MG4 C  | Battery=1  64kWh
MG4 L  | Battery=1  64kWh

ZS SR  | Battery=2  50,3kWh
ZS LR  | Battery=?  69,9kWh

Marvel r has a 70kWh battery, seems that this is the actual default value.

It looks like the lower value of battery is the higher capacity. With the model it should be possible to create a internal mapping table.

**Comment by tosate** [comment]
The owner manual of the ZS mentions battery packet type 1 and type 2:

_Im Durchschnitt dauert es mindestens 11,5 Stunden (Batteriepaket Typ 1, einphasige Stromversorgung) bzw. 8,5 Stunden (Batteriepaket Typ 1, dreiphasige Stromversorgung) oder 8,5 Stunden (Batteriepaket Typ 2, einphasige Stromversorgung) um einen vollständigen Ladezyklus ein- schließlich Ausgleichsladung durchzuführen._

My ZS SR provides in the configuration string `name:BATTERY,code:BATTERY,value:1`.

ZS SR | Battery=1 50,3kWh
ZS LR | Battery=2 69,9kWh

**Comment by Tonno87** [comment]
Okay, i meant the BType value... Sorry :)

In the configuration table the value of battery is allways 1:
![image](https://github.com/SAIC-iSmart-API/saic-python-mqtt-gateway/assets/59800400/6a5b1bbc-8163-430a-97f2-a0d30a96506a)
![image](https://github.com/SAIC-iSmart-API/saic-python-mqtt-gateway/assets/59800400/33e8776e-e0a3-4350-bc5b-c29585a6f075)

**Comment by nanomad** [comment]
@Tonno87 I've just discovered that BType is used as the feature gate for the the target battery percentage during scheduled charging
- BType 1 has it enabled and configurabile via the car
- Btype 2 has it disabled by the car and can be controller by an official MG chargers if BmsAdpPubChrgSttnDspCmd is 2

**Comment by Tonno87** [comment]
It's not possible to configure the charging percentage on cars with LFP battery (by car). So BType 2 is LFP in my opinion. 
Should work as long as MG won't release different NMC battery capacities per model.

**Comment by nanomad** [comment]
Unfortunately there's already a 77kwh mg4 so there we will have at least 3
capacities

Il Sab 5 Ago 2023, 12:15 Dennis Wehr ***@***.***> ha scritto:
